subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/processing/IncrementalFiler.java

```diff
@@ -16,41 +16,89 @@
 
-    IncrementalFiler(Filer delegate, Messager messager) {
+    IncrementalFiler(Filer delegate, AnnotationProcessingResult result, Messager messager) {
         this.delegate = delegate;
+        this.result = result;
         this.messager = messager;
     }
 
     @Override
     public final JavaFileObject createSourceFile(CharSequence name, Element... originatingElements) throws IOException {
-        checkOriginatingElements(name, originatingElements, messager);
+        recordGeneratedType(name, originatingElements, messager);
         return delegate.createSourceFile(name, originatingElements);
     }
 
     @Override
     public final JavaFileObject createClassFile(CharSequence name, Element... originatingElements) throws IOException {
-        checkOriginatingElements(name, originatingElements, messager);
+        recordGeneratedType(name, originatingElements, messager);
         return delegate.createClassFile(name, originatingElements);
     }
 
-    protected abstract void checkOriginatingElements(CharSequence name, Element[] originatingElements, Messager messager);
+    private void recordGeneratedType(CharSequence name, Element[] originatingElements, Messager messager) {
+        String generatedType = name.toString();
+        Set<String> originatingTypes = getTopLevelTypeNames(originatingElements);
+        checkGeneratedType(generatedType, originatingTypes, messager);
+        result.addGeneratedType(generatedType, originatingTypes);
+    }
+
+    protected abstract void checkGeneratedType(String generatedType, Set<String> orignatingTypes, Messager messager);
+
+    private Set<String> getTopLevelTypeNames(Element[] originatingElements) {
+        if (originatingElements == null || originatingElements.length == 0) {
+            return Collections.emptySet();
+        }
+        if (originatingElements.length == 1) {
+            String topLevelTypeName = getTopLevelTypeName(originatingElements[0]);
+            return topLevelTypeName != null ? Collections.singleton(topLevelTypeName) : Collections.<String>emptySet();
+        }
+        Set<String> typeNames = Sets.newLinkedHashSet();
+        for (Element element : originatingElements) {
+            String topLevelTypeName = getTopLevelTypeName(element);
+            if (topLevelTypeName != null) {
+                typeNames.add(topLevelTypeName);
+            }
+        }
+        return typeNames;
+    }
+
+    private String getTopLevelTypeName(Element originatingElement) {
+        Element current = originatingElement;
+        Element parent = originatingElement;
+        while (parent != null && !(parent instanceof PackageElement)) {
+            current = parent;
+            parent = current.getEnclosingElement();
+        }
+        if (!(current instanceof TypeElement)) {
+            messager.printMessage(Diagnostic.Kind.ERROR, "Incremental annotation processors must use types (or elements contained in types) as originating elements.");
```

Comment:
I don't think this should be an error - it's ok if users use some non-type elements to generate files. I think a lower message level should be used here.

Reply:
Oh sorry, we get here only when it's declared as incremental. Forget it.